### PR TITLE
Add table.product operator

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/theory/datatypes/datatypes_utils.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/******************************************************************************
```
Comment (ajreynol) [inline src/theory/datatypes/datatypes_utils.h:1]: Note there is already `theory_datatypes_utils.h`.

Although this file should perhaps be called `tuple_utils.h`.

File: src/theory/bags/bags_rewriter.cpp [hunk 2]
```diff
@@ -654,6 +655,20 @@ BagsRewriteResponse BagsRewriter::postRewriteFold(const TNode& n) const
   }
   return BagsRewriteResponse(n, Rewrite::NONE);
 }
+
+BagsRewriteResponse BagsRewriter::postRewriteProduct(const TNode& n) const
```
Comment (ajreynol) [inline src/theory/bags/bags_rewriter.cpp:659]: I'd expect there to be a rewrite that does evaluation in the case that all arguments are constant.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/datatypes/datatypes_utils.h (hunk 1 above)
Comment (mudathirmahgoub) [reply src/theory/datatypes/datatypes_utils.h]: File renamed to tuple_utils and class renamed to TupleUtils.

File: src/theory/bags/bags_rewriter.cpp (hunk 2 above)
Comment (mudathirmahgoub) [reply src/theory/bags/bags_rewriter.cpp:659]: Yes, this function is only called if not all children are constants. 
When they are constants, this function
https://github.com/cvc5/cvc5/blob/master/src/theory/bags/bags_rewriter.cpp#L68
will call `evaluateProduct` which computes the product from constants. 
https://github.com/cvc5/cvc5/pull/8020/files#diff-a9026cb0a61c5f5cc41cb5b0753617a74d133dd35186dc08fa37d192df4f69a2R119

Comment (ajreynol) [reply src/theory/bags/bags_rewriter.cpp:659]: Ah cool, didnt realize this was handled more generally.

File: test/regress/regress1/bags/product3.smt2 [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+(declare-fun x () (Tuple Int Int Int))
+(declare-fun y () (Tuple Int Int Int))
+(declare-fun z () (Tuple Int Int Int Int Int Int))
+
+(assert (bag.member x A))
+(assert (bag.member y B))
+(assert (bag.member z C))
+
+(assert (distinct x y ((_ tuple_project 0 1 2) z ((_ tuple_project 3 4 5) z))))
```
Comment (ajreynol) [inline test/regress/regress1/bags/product3.smt2:19]: Does this parse?  Seems like there is a missing paren after the first z.

Comment (mudathirmahgoub) [reply test/regress/regress1/bags/product3.smt2:19]: The parse didn't complain :) Do we have a problem in the parser?

Comment (ajreynol) [reply test/regress/regress1/bags/product3.smt2:19]: Yes, this is a problem (assuming you agree this is not what you meant to write).

Comment (mudathirmahgoub) [reply test/regress/regress1/bags/product3.smt2]: Fixed. I just put the parentheses in the wrong places. Both versions return sat (with different models)

File: src/theory/sets/rels_utils.h [hunk 4]
```diff
@@ -10,90 +10,60 @@
...
+ public:
+  /**
+   * compute the transitive closure of a binary relation
+   * @param members constant nodes of type (Tuple E E) that are known to in the
+   * relation rel
+   * @param rel a binary relation of type (Set (Tuple E E))
+   * @pre all members need to be constants
+   * @return the transitive closure of the relation
+   */
+  static std::set<Node> computeTC(std::set<Node> members, Node rel);
```
Comment (ajreynol) [inline src/theory/sets/rels_utils.h:36]: Thanks for updating this, let's update this to pass a reference. `std::set<Node>& members`.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/bags/bags_utils.h [hunk 5]
```diff
@@ -94,6 +94,30 @@ class BagsUtils
...
   static Node evaluateBagFilter(TNode n);
 
+  /**
+   * @param n of the form (table.product A B) where A , B of types (Bag T1),
+   * (Bag T2) respectively.
+   * @param e1 a tuple of type T1 of the form (tuple a1 ... an)
+   * @param e2 a tuple of type T2 of the form (tuple b1 ... bn)
+   * @return  (tuple a1 ... an b1 ... bn)
+   */
+  static Node constructTableTuple(TNode n, std::vector<TNode> elements);
```
[the remark this answers is not in the document]
Comment (mudathirmahgoub) [reply src/theory/bags/bags_utils.h:104]: Actually this function is replaced with TupleUtils::constructTupleFromElements which does receive a constant reference to a vector. 
So this function is removed.

File: src/theory/sets/rels_utils.h (hunk 4 above)
Comment (mudathirmahgoub) [reply src/theory/sets/rels_utils.h]: Done.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**